File: practise/algs/sorting_algs/radix.py
def counting_sort(arr, exp):
    n = len(arr)
    count_arr = [0] * 10

    for i in range(n):
        digit = arr[i] // exp % 10
        count_arr[digit] += 1

    for i in range(1, 10):
        count_arr[i] += count_arr[i - 1]

    out = [0] * n
    for i in range(n - 1, -1, -1):
        digit = arr[i] // exp % 10
        out[count_arr[digit] - 1] = arr[i]
        count_arr[digit] -= 1
    
    for i in range(n):
        arr[i] = out[i]

def radix_sort(arr):
    m = max(arr)

    exp = 1
    while m // exp > 0:
        counting_sort(arr, exp)
        exp *= 10

    print("Radix sort: ", end="")

File: practicec/posetim u snovime s vremena na vreme/radix.c
#include <stdio.h>

int maxElement(int* arr, int n);
void radixSort(int* arr, int n);

int main() {
    int arr[] = {4, 75, 123, 19, 183, 34};    
    int n = sizeof(arr) / sizeof(arr[0]);
    radixSort(arr, n);
    int i;
    for(i = 0; i < n; i++) printf("%d ", arr[i]);
    return 0;
}

int maxElement(int* arr, int n)
{
    int max = arr[0];
    int i;
    
    for(i = 1; i < n; i++)
    {
        if(arr[i] > max) max = arr[i];
    }

    return max;
}

void radixSort(int* arr, int n)
{
    int exp = 1;
    int max = maxElement(arr, n);
    int digitsCount[10];
    int out[n];
    int i;

    while(max / exp > 0)
    {
        for(i = 0; i < 10; i++) digitsCount[i] = 0;
        for(i = 0; i < n; i++)
        {
            digitsCount[arr[i] / exp % 10]++;
        }
        for(i = 1; i < 10; i++) digitsCount[i] += digitsCount[i - 1];
        for(i = n - 1; i >= 0; i--)
        {
            out[--digitsCount[arr[i] / exp % 10]] = arr[i];
        }
        for(i = 0; i < n; i++) arr[i] = out[i];
        exp *= 10;
    }
}